src/v/iceberg/tests/schema_protobuf_test.cc

```diff
@@ -0,0 +1,187 @@
+  const ss::sstring& proto_str, google::protobuf::DescriptorPool& pool) {
+    google::protobuf::compiler::Parser parser;
+
+    google::protobuf::io::ArrayInputStream input_stream(
+      proto_str.c_str(), proto_str.size());
+
+    google::protobuf::io::Tokenizer tokenizer(&input_stream, nullptr);
+    google::protobuf::FileDescriptorProto fdp;
+    fdp.set_name("test_schema_file");
+
+    auto parse_result = parser.Parse(&tokenizer, &fdp);
+    ASSERT_TRUE(parse_result);
+    pool.BuildFile(fdp);
+}
+
+}; // namespace
+
+MATCHER_P3(IsField, id, name, type, "") {
+    fmt::print(
+      "field(id: {}, name: {}, type: {}) expected to be equal to field(id: {}, "
+      "name: {}, type: {})\n",
+      arg->id,
+      arg->name,
+      arg->type,
+      id,
+      name,
+      type);
+    *result_listener << "";
+
+    return arg->id == id && arg->name == name && arg->type == type;
+}
+
+TEST(SchemaProtobuf, TestSimpleSchemaConversion) {
+    google::protobuf::DescriptorPool d_pool;
+    protobuf_schema_converter converter;
+    parse_proto_definition_string(basic_proto_str, d_pool);
+    auto person_descriptor = d_pool.FindMessageTypeByName("Person");
+
+    auto result = converter.to_iceberg(*person_descriptor);
+    ASSERT_FALSE(result.is_error());
+    auto& top_level = result.value();
+    // string name = 1;
+    EXPECT_THAT(top_level.fields[0], IsField(0, "name", string_type{}));
+    // string id = 2;
+    EXPECT_THAT(top_level.fields[1], IsField(1, "id", int_type{}));
+    // string email = 3;
+    EXPECT_THAT(top_level.fields[2], IsField(2, "email", string_type{}));
+
+    // Department dept = 4; (nested struct)
+    EXPECT_EQ(top_level.fields[3]->id, 3);
+    EXPECT_EQ(top_level.fields[3]->name, "dept");
+    auto& nested = std::get<struct_type>(top_level.fields[3]->type);
+    // int32 id = 1;
+    // string name = 2;
+    EXPECT_THAT(nested.fields[0], IsField(0, "id", int_type{}));
+    EXPECT_THAT(nested.fields[1], IsField(1, "name", string_type{}));
+    // string test_coverage = 5;
+    EXPECT_THAT(
+      top_level.fields[4], IsField(4, "test_coverage", string_type{}));
+}
+
+/**
+ *enum State {
+ *  Active = 0;
+ *  InActive = 1;
+ *}
+ *
+ *message BrokerShard {
+ *  int64 id = 1;
+ *  int32 shard = 2;
+ *}
+ *
+ *message Partition {
+ *    int32 id = 1;
+ *    repeated BrokerShard replicas = 2;
+ *    State state = 3;
+ *}
+ *
+ *message Topic {
+ *    string name = 1;
+ *    repeated Partition partitions = 2;
+ *}
+ *
+ *message Metadata {
+ *    map<string,Topic> topics = 1;
+ *    repeated int64 nodes = 2;
+ *}
+ */
+// this schema contains a list map and nested fields
+static_assert(
+  SchemaConverter<protobuf_schema_converter, google::protobuf::Descriptor>);
+TEST(SchemaProtobuf, TestComplexSchema) {
+    google::protobuf::DescriptorPool d_pool;
+    protobuf_schema_converter converter;
+    parse_proto_definition_string(complex_proto_str, d_pool);
+    auto md = d_pool.FindMessageTypeByName("Metadata");
```

Comment:
This is an interesting case. In the case where there are nested PB schemas like this, we'll likely need the catalog/coordinator assign a mapping from name to Iceberg field id, and then pass that mapping to the converter. As I understand it, the top level table will need unique field ids.

Since the first cut of support will not have schema evolution, we might even be able to have the converter assign unique field ids, and worry about new assignments later on.. though that probably needs a bit more thought.

I think it's worth proceeding with this PR as you've done with idx==field_id, but just something to be aware of.